### DISCUSSION TITLE
[FEATURE REQUEST] - vis and admin "keystroke"

### DISCUSSION THREAD

**Discussion by hiranokite** [moudey/Shell#181] [Ideas]
can we have a option to make "vis" and "admin" work without need to keep the activation key pressed until the context menu show-up or the app start?

to show hidden menus i prefer to hold LMB and press RMB, but i can use Shift Keys and press RMB too
`	menu(type='desktop' title='Shor&tcuts' separator="before" image=\uE14A vis=@key(1) or @key(key.lshift) or @key(key.rshift))`
and
to open file as admin i prefer to hold RMB and press LMB, but i can use Shift Keys and press LMB too
`	item(title='Open with Notepad'type='file' admin=@key(2) or @key(key.lshift) or @key(key.rshift) cmd='notepad.exe' args='@sel.path.quote' image='notepad.exe')`

**Comment by moudey** [comment]
The keyboard has two events, the first before the menu appears, and the second before the command is executed

`RMB` + `LMB` = `CTRL` + `RMB`

**Comment by moudey** [comment]
Refer to the terminal file, you will find that the value of `true` is assigned to the `admin` property in the event that <kbd>Shift</kbd> only before executing the command and not before the menu appears.

**Comment by hiranokite** [comment]
Sorry, my English is not the best, here is a example of what I'm trying to explain.

![image](https://user-images.githubusercontent.com/5361459/224472319-f2661e41-706c-4ebe-8294-f8a52b2f6b04.png)

the last delay of `50ms` is the problem, `with 50ms everything works perfectly` but, if you lower the number to `30ms or less` the key activation will not work. so this mean you need to keep the activation key pressed til the menu open.
in my use I hold shift and right after press the right click I release the shift so the `vis=key.shift()` take no effect

**Comment by moudey** [comment]
It works fine with me. I will check this issue.

Mouse functionality has been added to the command execution event. Where you can run the command as administrator, for example, by right-clicking on the item
```
item(title="Command Prompt" admin=key.rbutton()  cmd args='/K echo  RB=@key.rbutton(), LB=@key.lbutton()')
```
https://nilesoft.org/download/shell/debug.zip

**Reply by hiranokite** [reply]
maybe this delay is a windows side delay, since it exists too in default context menu.
tried set delay to 0 on this windows key, but no success.

`HKEY_CURRENT_USER\Control Panel\Desktop`
`MenuShowDelay`

i cant release shift until the menu appears.


> Mouse functionality has been added to the command execution event. Where you can run the command as administrator, for example, by right-clicking on the item
> 
> ```
> item(title="Command Prompt" admin=key.rbutton()  cmd args='/K echo  RB=@key.rbutton(), LB=@key.lbutton()')

nice, i was using the default keys values, `LB=@key(1)` and `RB=@key(2)`, updated them to match the new syntax.

since you mentioned this, there is a bug on taskbar utilizing LB. with shift this not occurs, I'll try to explain.

on taskbar.nss change the fisrt line to this
`menu(type="taskbar" vis=key.lbutton() or key.shift() pos=0 title=app.name image=\uE249)`
reload shell `😜`
now you can use LB > RB to show the hidden shell menu

Follow these steps!
1. RB on taskbar open context menu.
2. hold LB and press RB will open the context menu with the hidden shell menu.
3. you can hold LB and press RB many times you want it will open the context menu with the hidden shell menu without problem.
4. but if you do the double click or more on taskbar with LB you cant open the context menu with hidden shell menu with LB > RB until you click one time with RB on taskbar.

This bug occurs on older versions too but using the `@key(1)` instead of `@key.lbutton()`, with means the same.

**Reply by moudey** [reply]
>MenuShowDelay
```
set {
    showdelay=0
}
```

>RB on taskbar open context menu.

RB has been activated to execute the command on the taskbar

>but if you do the double click or more on taskbar with LB you cant open the context menu with hidden shell menu with LB > RB until you click one time with RB on taskbar.

This issue has been fixed

Try this [build](https://nilesoft.org/download/shell/debug.zip)

**Reply by hiranokite** [reply]
> > MenuShowDelay
> 
> ```
> set {
>     showdelay=0
> }
> ```
> 

I tried this too to fix the initial delay but its only affect the sub-menus on hover

> > RB on taskbar open context menu.
> 
> RB has been activated to execute the command on the taskbar
> 
> > but if you do the double click or more on taskbar with LB you cant open the context menu with hidden shell menu with LB > RB until you click one time with RB on taskbar.
> 
> This issue has been fixed
> 
> Try this [build](https://nilesoft.org/download/shell/debug.zip)

Nice, tested this new build and worked fine.

there's another old bug on taskbar using multiple monitors I see some users reporting this but I think the real reason has not reported.
This only happens on a non-primary monitor, and cause the shell to not hook into taskbar, so it show the default Windows 11 context menu.

![image](https://user-images.githubusercontent.com/5361459/225808054-73c26ffb-d265-4b84-b8cf-c0a90b586d4b.png)

so this bug occurs on those occasions that I noticed.

on secondary monitor
- Turn off and on monitor
- New monitor plugged-in
- Monitor or PC enter sleep mode

**Reply by moudey** [reply]
>on secondary monitor
>- Turn off and on monitor
>- New monitor plugged-in
>- Monitor or PC enter sleep mode

This issue has been fixed with this [build](https://nilesoft.org/download/shell/debug.zip)

**Reply by hiranokite** [reply]
> > on secondary monitor
> > 
> > * Turn off and on monitor
> > * New monitor plugged-in
> > * Monitor or PC enter sleep mode
> 
> This issue has been fixed with this [build](https://nilesoft.org/download/shell/debug.zip)

wow that was fast thanks, tested and everything seems working great.

I found a bug on Explorer titlebar.

- Shell on titlebar only show when invoked by the mouse.
- If you open a new explorer and press Alt+Space the titlebar context menu will show without Shell.

![bug11-1](https://user-images.githubusercontent.com/5361459/226077629-8d0ac0ff-6c4a-4159-8e4b-8692d198a13b.png)

- and once you right click on titlebar to show the Shell if you press again the shortcut Alt+Space, the titlebar context menu will show like this.

![bug11-2](https://user-images.githubusercontent.com/5361459/226077686-d4142b20-a7ae-49a2-bfc2-31b8f2ee014a.png)

**Reply by moudey** [reply]
This menu does not use the `TrackPopupMenu` function, so it cannot be dealt with, because it will need to use monitor all windows, and this will affect performance, Because I always try to use light methods that do not affect performance

#118

**Reply by hiranokite** [reply]
> This menu does not use the `TrackPopupMenu` function, so it cannot be dealt with, because it will need to use monitor all windows, and this will affect performance, Because I always try to use light methods that do not affect performance
> 
> #118

I see, I just found it odd that the Alt+Space menu bugs once Shell titlebar is invoked.
but since this menu almost never used i prefer performance too.

**Reply by hiranokite** [reply]
> > on secondary monitor
> > 
> > * Turn off and on monitor
> > * New monitor plugged-in
> > * Monitor or PC enter sleep mode
> 

found that if you change the primary monitor, the old primary monitor taskbar will not show the Shell context menu.

**Reply by moudey** [reply]
>found that if you change the primary monitor, the old primary monitor taskbar will not show the Shell context menu.

> and once you right click on titlebar to show the Shell if you press again the shortcut Alt+Space, the titlebar context menu will show like this.

These issues have been fixed with this [build](https://nilesoft.org/download/shell/debug.zip)

**Reply by hiranokite** [reply]
> > found that if you change the primary monitor, the old primary monitor taskbar will not show the Shell context menu.
> 
> > and once you right click on titlebar to show the Shell if you press again the shortcut Alt+Space, the titlebar context menu will show like this.
> 
> These issues have been fixed with this [build](https://nilesoft.org/download/shell/debug.zip)

did you tested this new build on Windows 11? on my pc only fixed if i use with [ExplorerPatcher](https://github.com/valinet/ExplorerPatcher)

**Reply by moudey** [reply]
>did you tested this new build on Windows 11? on my pc only fixed if i use with [ExplorerPatcher](https://github.com/valinet/ExplorerPatcher)

Tested on Windows 11 with/without ExplorerPatcher

Try it with Windows Sandbox

**Reply by hiranokite** [reply]
tested on a VM, still the same bug

![image](https://user-images.githubusercontent.com/5361459/226482400-ac050708-c32b-4bac-87cb-86d430f40386.png)

**Reply by moudey** [reply]
Try this [build](https://nilesoft.org/download/shell/debug.zip)

**Reply by hiranokite** [reply]
> Try this [build](https://nilesoft.org/download/shell/debug.zip)

not fixed, and now with ExplorerPatcher has a new bug.

**Reply by moudey** [reply]
Please explain more

**Reply by hiranokite** [reply]
I recorded some videos of the VM while testing.

1. Windows 11 + Shell
After changing primary monitor, old primary no longer has Shell in taskbar.
https://cdn.discordapp.com/attachments/894086568330027008/1087880527685636107/2023-03-21_20-22-33.mp4

2. Windows 11 + Shell + ExplorerPatcher
Here is the new bug with this build, it's even hard to explain, better watch the video, taskbar even thinks its Windows 10 Taskbar showing Windows 10 taskbar menus but everything mess.
https://cdn.discordapp.com/attachments/894086568330027008/1087880652411637852/2023-03-21_20-23-56.mp4

4. Windows 11 + Shell + ExplorerPatcher with Windows 10 Taskbar
I didn't notice any bug using the Windows 10 taskbar, everything seems to work normally.
https://cdn.discordapp.com/attachments/894086568330027008/1087880768895844433/2023-03-21_20-27-12.mp4

**Reply by hiranokite** [reply]
> Please explain more

Sorry, I was just finishing the tests.

**Reply by moudey** [reply]
@hiranokite Test it

**Reply by hiranokite** [reply]
> @hiranokite Test it

seems fixed, but the old bug has "returned"

did you able to reproduce the bug on your PC?

**Reply by moudey** [reply]
> but the old bug has "returned"
did you able to reproduce the bug on your PC?

- Tested with switching the primary display
- Tested with multiple displays
- Tested ExplorerPatcher
- Not tested ExplorerPatcher with switching the main display and multiple displays

**Reply by moudey** [reply]
How to combine displays in a video?

**Reply by hiranokite** [reply]
ok, so test turning off and on the primary monitor.

**Reply by hiranokite** [reply]
> How to combine displays in a video?

that was a VM i just expanded the screens and cropped the record area.

**Reply by moudey** [reply]
Sorry test it now

**Reply by hiranokite** [reply]
> Sorry test it now

Tested so many combinations, seems unbreakable now!
worked on every test with and without ExplorerPatcher.

**Reply by moudey** [reply]
This [build](https://nilesoft.org/download/shell/debug.zip) works on all connected displays

**Reply by hiranokite** [reply]
> This [build](https://nilesoft.org/download/shell/debug.zip) works on all connected displays

Nice, since I only have 2 displays I can't test it properly but thanks.
on my tests everything worked flawless on Windows 11
not tested on Windows 10 or 11 with 10 taskbar (ExplorerPatcher).

**Comment by hiranokite** [comment]
if possible can you add a button on shell to open the shell.exe location on explorer, or maybe just a hidden shortcut eg. `Crtl+S` or change email to Ctrl+M and add Ctrl+E to open shell on explorer, it will be helpful while testing.

![633CB9CC5C965337589A1EA1E5C4FCF5](https://user-images.githubusercontent.com/5361459/224620161-e4a0cc0e-8e35-4cda-a2d4-f5310aa916d5.png)

**Comment by hiranokite** [comment]
@moudey seems the new build broke taskbar context menu, are you aware of this?
can't get to work with my or default nss files.
Windows 11 build 22621.1485

Only show the default windows menu.

**Reply by moudey** [reply]
It was working fine on Windows, but it didn't work on Sandbox. Now it works
